packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/include/camera_avfoundation/FLTCaptureDevice.h

```diff
@@ -63,17 +71,41 @@ NS_ASSUME_NONNULL_BEGIN
 - (CMTime)activeVideoMaxFrameDuration;
 - (void)setActiveVideoMaxFrameDuration:(CMTime)duration;
 
-- (AVCaptureInput *)createInput:(NSError *_Nullable *_Nullable)error;
+@end
+
+/// A protocol which is a direct passthrough to AVCaptureInput.
+/// It exists to allow replacing AVCaptureInput in tests.
+@protocol FLTCaptureInput <NSObject>
+
+/// Underlying input instance. It is exposed as raw AVCaptureInput has to be passed to some
+/// AVFoundation methods. The plugin implementation code shouldn't use it though.
+@property(nonatomic, readonly) AVCaptureInput *input;
 
+@property(nonatomic, readonly) NSArray<AVCaptureInputPort *> *ports;
 @end
 
-/// A default implementation of FLTCaptureDeviceControlling protocol which
-/// wraps an instance of AVCaptureDevice.
-@interface FLTDefaultCaptureDeviceController : NSObject <FLTCaptureDeviceControlling>
+/// A protocol which wraps the creation of AVCaptureDeviceInput.
+/// It exists to allow mocking instances of AVCaptureDeviceInput in tests.
+@protocol FLTCaptureDeviceInputFactory <NSObject>
+- (nullable NSObject<FLTCaptureInput> *)deviceInputWithDevice:(NSObject<FLTCaptureDevice> *)device
+                                                        error:(NSError **)error;
+@end
 
-/// Initializes the controller with the given device.
+/// A default implementation of `FLTCaptureDevice` which is a direct passthrough to the underlying
+/// `AVCaptureDevice`.
+@interface FLTDefaultCaptureDevice : NSObject <FLTCaptureDevice>
```

Comment:
why do we need the wrapper here? Can we directly have `AVCaptureDevice` conform to `FLTCaptureDevice` protocol?

Reply:
Please refer to the above discussion with @stuartmorgan https://github.com/flutter/packages/pull/8363#discussion_r1934276991

Reply:
am i reading it right that the reason here is that some AVFoundation APIs expect the actual `AVCaptureDevice` type to be passed in?

Reply:
That is related, but the biggest problem is that categories in Obj-C don't provide good type safety. In some cases we might be passing wrong type and the compiler won't warn us.